spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,474 @@
+ * level merging of Parquet files when applicable.
+ *
+ * <p>This runner uses {@link ParquetFileMerger} to merge Parquet files at the row-group level
+ * without full deserialization, which is significantly faster than the standard Spark rewrite
+ * approach for Parquet files.
+ *
+ * <p>The decision to use this runner vs. SparkBinPackFileRewriteRunner is controlled by the
+ * configuration option {@code use-parquet-file-merger}.
+ */
+public class SparkParquetFileMergeRunner extends SparkBinPackFileRewriteRunner {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkParquetFileMergeRunner.class);
+  private final FileRewriteCoordinator coordinator = FileRewriteCoordinator.get();
+
+  public SparkParquetFileMergeRunner(SparkSession spark, Table table) {
+    super(spark, table);
+  }
+
+  @Override
+  public String description() {
+    return "PARQUET-MERGE";
+  }
+
+  @Override
+  protected void doRewrite(String groupId, RewriteFileGroup group) {
+    // Early validation: check if requirements are met and get schema
+    MessageType schema = validateAndGetSchema(group);
+    if (schema == null) {
+      LOG.info(
+          "Row-group merge requirements not met for group {}. Using standard Spark rewrite.",
+          groupId);
+      super.doRewrite(groupId, group);
+      return;
+    }
+
+    // Requirements met - attempt row-group level merge
+    try {
+      LOG.info(
+          "Merging {} Parquet files using row-group level merge (group: {})",
+          group.rewrittenFiles().size(),
+          groupId);
+      mergeParquetFilesDistributed(groupId, group, schema);
+    } catch (Exception e) {
+      LOG.warn(
+          "Row-group merge failed for group {}, falling back to standard Spark rewrite: {}",
+          groupId,
+          e.getMessage(),
+          e);
+      // Fallback to standard rewrite
+      super.doRewrite(groupId, group);
+    }
+  }
+
+  /**
+   * Validates if the file group can use row-group level merging and returns the schema.
+   *
+   * <p>Requirements:
+   *
+   * <ul>
+   *   <li>All files must be Parquet format
+   *   <li>Table must not have a sort order (no sorting or z-ordering)
+   *   <li>Files must not have delete files or delete vectors
+   *   <li>Files must have compatible schemas (verified by ParquetFileMerger.readAndValidateSchema)
+   *   <li>Files must not be encrypted (detected by ParquetCryptoRuntimeException)
+   * </ul>
+   *
+   * @param group the file group to check
+   * @return the Parquet schema if row-group merging can be used, null otherwise
+   */
+  private MessageType validateAndGetSchema(RewriteFileGroup group) {
+    // Check if all files are Parquet format
+    boolean allParquet =
+        group.rewrittenFiles().stream().allMatch(file -> file.format() == FileFormat.PARQUET);
+
+    if (!allParquet) {
+      LOG.debug("Cannot use row-group merge: not all files are Parquet format");
+      return null;
+    }
```

Comment:
Should be part of the `ParquetFileMerger` check

Reply:
good catch